### ISSUE TITLE
Fix lowdb build

### ISSUE THREAD

**Issue by aj-r** [DefinitelyTyped/DefinitelyTyped#25002]
`lowdb-tests` are failing on master, probably due to a typescript change. It looks like the `If<...>` types no longer work properly.

The good news is, in TS2.8 we can use conditional types, which do the same thing in a much more reliable way. I can make a PR for this, but I want to make sure we're OK with dropping support for TS 2.6-2.7.

I think it's worth dropping support because the types are very broken in TS2.8 right now. Anyone using TS 2.6 can still install `@types/lowdb@ts2.6` to pin to the older version.

- [x] I tried using the `@types/xxxx` package and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [x] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [x] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @typicode @niieani

**Comment by niieani** [comment]
If you're up for making the PR, go for it 👍 I don't have the time to do it myself at this moment, but I think conditional types will really simplify the current code. 

If you need any help or context, let me know.

**Comment by aj-r** [comment]
tl;dr I think we need to write (or auto-generate) our own version of `LoDashExplicitWrapper`. Apologies for the massive comment.

This is a problem: DefinitelyTyped won't let us target TS 2.8. I'm not sure if this is intentional or if `dtslint` just needs to be updated.

However I'm finding that conditional types don't solve all of our problems anyways. The strategy of recursively tagging objects doesn't work very well: typescript fails to compile if you want to have a recursive array type. Additionally, tagging doesn't work (even the current way we do tagging) if the user does something like this:

```ts
db.get("posts")
  .find({ id: 123 })
  .map(post => post.id.toLowerCase()) // calling toLowerCase() returns an untagged string, so write() won't work properly
  .write();
```

A better solution, which also reflects what lowdb is really doing, would be to create our own  lodash wrapper types. I tried to do this using conditional types, but unfortunately they don't work very well when modifying functions with generic type parameters. My best attempt was this:

```ts
export type LowdbExplicitSyncWrapper<TValue> = {
    [P in keyof LoDashExplicitWrapper<TValue> | "write"]:
      P extends "value" | "valueOf" | "toJSON" ? LoDashExplicitWrapper<TValue>[P] :
      P extends keyof LoDashExplicitWrapper<TValue> ? (
        LoDashExplicitWrapper<TValue>[P] extends <T>(arg: infer T1) => LoDashExplicitWrapper<infer R> ? <T>(arg: T1) => LowdbExplicitSyncWrapper<R> :
        LoDashExplicitWrapper<TValue>[P]) :
      () => TValue;
  }
```

However this results in a type like this:

```ts
type Lowdb.LowdbExplicitSyncWrapper<TValue> = {
    pop: (arg: {}) => LowdbExplicitSyncWrapper;
    push: (arg: {}) => LowdbExplicitSyncWrapper;
    ...
}
```

So the only solution seems to be rewriting the entire `LoDashExplicitWrapper`. Twice - once for sync and once for async. This clearly sucks, but there are some benefits:
- Types actually work in TS2.8
- No more dependency on `lodash`. I've found that when a `@types` package depends on another `@types` package, it can cause strange and annoying behaviors sometimes. So removing that dependency is a big plus.
- No more polluting models with tags that don't actually exist at runtime
- I can write a script to auto-generate our `LoDashExplicitWrapper` types based on lodash's version. This makes it easy to keep our version in sync with lodash.

@niieani thoughts?

**Comment by niieani** [comment]
Yes, I think that's a good solution. In the future, perhaps TypeScript should allow "deep copying" of types - that would solve our case here without code duplication.

@mhegazy Is there a way to add a hook to the DefinitelyTyped so that the wrapper is generated every time there's a new version of `lodash`?
